piker/brokers/kucoin.py

```diff
@@ -0,0 +1,668 @@
+                    },
+                    'shm_write_opts': {'sum_tick_vml': False},
+                    'fqsn': sym,
+                },
+            }
+
+            @acm
+            async def subscribe(ws: wsproto.WSConnection):
+                @acm
+                async def open_ping_task(ws: wsproto.WSConnection):
+                    '''
+                    Spawn a non-blocking task that pings the ws
+                    server every ping_interval so Kucoin doesn't drop
+                    our connection
+
+                    '''
+                    async with trio.open_nursery() as n:
+                        # TODO: cache this task so it's only called once
+                        async def ping_server():
+                            while True:
+                                await trio.sleep((ping_interval - 1000) / 1000)
+                                await ws.send_msg({'id': connect_id, 'type': 'ping'})
+
+                        log.info(f'Starting ping task for {sym}')
+                        n.start_soon(ping_server)
+
+                        yield ws
+
+                        n.cancel_scope.cancel()
+
+                # Spawn the ping task here
+                async with open_ping_task(ws) as ws:
+                    tasks = []
+                    tasks.append(make_sub(kucoin_sym, connect_id, level='l3'))
+                    tasks.append(make_sub(kucoin_sym, connect_id, level='l1'))
+
+                    for task in tasks:
+                        log.info(f'Subscribing to {task["topic"]} feed for {sym}')
+                        await ws.send_msg(task)
+
+                    yield
+
+                    # unsub
+                    if ws.connected():
+                        log.info(f'Unsubscribing to {kucoin_sym} feed')
+                        await ws.send_msg(
+                            {
+                                'id': connect_id,
+                                'type': 'unsubscribe',
+                                'topic': f'/market/ticker:{sym}',
+                                'privateChannel': False,
+                                'response': True,
+                            }
+                        )
+
+            async with open_autorecon_ws(
+                f'wss://ws-api-spot.kucoin.com/?token={token}&[connectId={connect_id}]',
+                fixture=subscribe,
+            ) as ws:
+                msg_gen = stream_messages(ws, sym)
+                typ, quote = await msg_gen.__anext__()
+
+                while typ != 'trade':
+                    # TODO: use ``anext()`` when it lands in 3.10!
+                    typ, quote = await msg_gen.__anext__()
+
+                task_status.started((init_msgs, quote))
+                feed_is_live.set()
+
+                async for typ, msg in msg_gen:
+                    await send_chan.send({sym: msg})
+
+
+def make_sub(sym, connect_id, level='l1') -> dict[str, str | bool]:
+    match level:
+        case 'l1':
```

Comment:
Yeah so re my comment about the trade rates mis-calculation, I think this might be the cause.

maybe one of these subs is being reported every 10ms and you're not filtering / diffing for actual updates since the last msg?

likely you'll need to keep a local book state and only transmit to piker when there's a real update to the book and/or real clear events that have taken place.

Reply:
yeah comment is just above https://github.com/pikers/piker/pull/494#issuecomment-1493068034